lib/tree.rb

```diff
@@ -16,40 +16,137 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(log n) if it's balanced; worst case is o(n)  
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    
+    if @root.nil?
+      @root = new_node
+      return
+    end
+
+    return add_helper(@root, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
+  end
+
+  # Time Complexity: o(log n) if it's balanced; worst case is o(n)
+  # Space Complexity: o(n)
   def find(key)
```

Comment:
👍  Space & Time complexities are both O(log n) unless the tree is unbalanced, in which case they are O(n).